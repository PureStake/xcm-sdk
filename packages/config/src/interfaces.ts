// eslint-disable-next-line import/no-cycle
import { Asset, Chain, MoonChain } from './constants';

export interface AssetConfig<Assets extends Asset = Asset> {
  id: string;
  /**
   * id -> erc20Id
   * `0xffffffff${BigInt(id).toString(16).padStart(32, '0')}`
   * exceptions are native tokens, for them we use
   * Erc20BalancesPrecompile 0x0000000000000000000000000000000000000802
   */
  erc20Id: string;
  originSymbol: Assets;
  originAssetId?: number;
  isNative?: boolean;
}

export interface ChainBaseConfig<Chains = Chain> {
  chain: Chains;
  name: string;
  ws: string;
  parachainId: number;
}

export interface MoonChainConfig extends ChainBaseConfig<MoonChain> {
  decimals: number;
}

export interface ChainConfig extends ChainBaseConfig {
  weight: number;
  moonAssetId?: number | bigint;
  palletInstance?: number;
}
