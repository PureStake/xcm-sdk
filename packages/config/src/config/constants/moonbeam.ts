import { createBalanceBuilder } from '../../balance';
import {
  Assets,
  Chain,
  MOONBEAM_ASSETS_CONFIGS as assets,
  MOONBEAM_CHINS_CONFIGS as chains,
  MoonChain,
  MOON_CHINS_CONFIGS,
} from '../../constants';
import {
  createExtrinsicBuilder,
  XTokensExtrinsicSuccessEvent,
} from '../../extrinsic';
import { MoonbeamAssets } from '../../interfaces';
import { createWithdrawBuilder } from '../../withdraw';
import { ChainXcmConfigs } from '../config.interfaces';

const config = MOON_CHINS_CONFIGS[MoonChain.Moonbeam];
const balance = createBalanceBuilder<MoonbeamAssets>();
const extrinsic = createExtrinsicBuilder<MoonbeamAssets>(config);
const withdraw = createWithdrawBuilder<MoonbeamAssets>();

export const MOONBEAM_CONFIGS: ChainXcmConfigs<MoonbeamAssets> = {
  [Assets.ACA]: {
    asset: assets[Assets.ACA],
    origin: chains[Chain.Acala],
    deposit: [
      {
        balance: balance.system(),
        extrinsic: extrinsic
          .xTokens()
          .transfer()
          .successEvent(XTokensExtrinsicSuccessEvent.TransferredMultiAssets)
          .params(chains[Chain.Acala], {
            Token: assets[Assets.ACA].originSymbol,
          }),
        origin: chains[Chain.Acala],
      },
    ],
    withdraw: [
      withdraw.xTokens({
        balance: balance.system(),
        destination: chains[Chain.Acala],
        existentialDeposit: 100_000_000_000,
        feePerWeight: 8,
      }),
    ],
  },
  [Assets.AUSD]: {
    asset: assets[Assets.AUSD],
    origin: chains[Chain.Acala],
    deposit: [],
    withdraw: [],
  },
  [Assets.DOT]: {
    asset: assets[Assets.DOT],
    origin: chains[Chain.Polkadot],
    deposit: [],
    withdraw: [],
  },
  [Assets.GLMR]: {
    asset: assets[Assets.GLMR],
    origin: config,
    deposit: [],
    withdraw: [],
  },
  [Assets.PARA]: {
    asset: assets[Assets.PARA],
    origin: chains[Chain.Parallel],
    deposit: [],
    withdraw: [],
  },
};
