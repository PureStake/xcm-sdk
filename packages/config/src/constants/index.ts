/* eslint-disable import/no-cycle */
export * from './assets';
export * from './chains';
