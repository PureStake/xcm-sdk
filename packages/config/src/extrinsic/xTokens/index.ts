export * from './xTokens';
export * from './xTokens.constants';
export * from './xTokens.interfaces';
