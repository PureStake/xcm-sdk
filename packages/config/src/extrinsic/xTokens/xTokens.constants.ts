export enum XTokensExtrinsic {
  Transfer = 'transfer',
  TransferMultiAsset = 'transferMultiasset',
  TransferMultiCurrencies = 'transferMulticurrencies',
}

export enum XTokensExtrinsicSuccessEvent {
  Transferred = 'Transferred',
  TransferredMultiAssets = 'TransferredMultiAssets',
  TransferredMultiCurrencies = 'TransferredMultiCurrencies',
}

export enum XTokensExtrinsicCurrencyTypes {
  ForeignAsset = 'ForeignAsset',
  FungibleToken = 'FungibleToken',
  MantaCurrency = 'MantaCurrency',
  MiningResource = 'MiningResource',
  Native = 'Native',
  NativeToken = 'NativeToken',
  OtherReserve = 'OtherReserve',
  Token = 'Token',
}
