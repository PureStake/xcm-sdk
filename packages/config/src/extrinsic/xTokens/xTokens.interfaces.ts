import { AssetSymbol } from '../../constants';
import { AssetId } from '../../interfaces';
import {
  XTokensExtrinsic,
  XTokensExtrinsicCurrencyTypes,
  XTokensExtrinsicSuccessEvent,
} from './xTokens.constants';

export type XTokensPallet<Symbols extends AssetSymbol = AssetSymbol> =
  | XTokensTransferExtrinsic<Symbols>
  | XTokensTransferMultiAssetExtrinsic<Symbols>
  | XTokensTransferMultiCurrenciesExtrinsic<Symbols>;

export type XTokensTransferExtrinsic<
  Symbols extends AssetSymbol = AssetSymbol,
> = XTokensBaseExtrinsic<XTokensExtrinsic.Transfer, Symbols>;
export type XTokensTransferMultiAssetExtrinsic<
  Symbols extends AssetSymbol = AssetSymbol,
> = XTokensBaseExtrinsic<XTokensExtrinsic.TransferMultiAsset, Symbols>;
export type XTokensTransferMultiCurrenciesExtrinsic<
  Symbols extends AssetSymbol = AssetSymbol,
> = XTokensBaseExtrinsic<XTokensExtrinsic.TransferMultiCurrencies, Symbols>;

export interface XTokensBaseExtrinsic<
  Extrinsic extends XTokensExtrinsic,
  Symbols extends AssetSymbol = AssetSymbol,
> {
  pallet: 'xTokens';
  extrinsic: Extrinsic;
  successEvent: XTokensExtrinsicSuccessEvent;
  getParams: (
    account: string,
    amount: bigint,
    fee?: bigint,
  ) => XTokensParamsByExtrinsic<Symbols>[Extrinsic];
}

export interface XTokensParamsByExtrinsic<
  Symbols extends AssetSymbol = AssetSymbol,
> {
  [XTokensExtrinsic.Transfer]: XTokensTransferExtrinsicParams<Symbols>;
  [XTokensExtrinsic.TransferMultiAsset]: XTokensTransferMultiAssetExtrinsicParams<Symbols>;
  [XTokensExtrinsic.TransferMultiCurrencies]: XTokensTransferMultiCurrenciesExtrinsicParams<Symbols>;
}

export type XTokensTransferExtrinsicParams<
  Symbols extends AssetSymbol = AssetSymbol,
> = [
  /**
   * asset
   */
  XTokensTransferExtrinsicParamsAsset<Symbols>,
  /**
   * amount
   */
  bigint,
  /**
   * destination
   */
  {
    V1: {
      parents: 1;
      interior: {
        X2: [
          {
            Parachain: number;
          },
          {
            AccountKey20: {
              network: 'Any';
              /**
               * account
               */
              key: string;
            };
          },
        ];
      };
    };
  },
  /**
   * weight
   */
  number,
];

export type XTokensTransferMultiAssetExtrinsicParams<
  Symbols extends AssetSymbol = AssetSymbol,
> = [
  /**
   * asset
   */
  {
    V1: {
      id: {
        Concrete: {
          parents: 1;
          interior: {
            X2: [
              {
                Parachain: number;
              },
              {
                GeneralKey: Symbols;
              },
            ];
          };
        };
      };
      fun: {
        /**
         * amount
         */
        Fungible: bigint;
      };
    };
  },
  /**
   * destination
   */
  {
    V1: {
      parents: 1;
      interior: {
        X2: [
          {
            Parachain: number;
          },
          {
            AccountKey20: {
              network: 'Any';
              key: string;
            };
          },
        ];
      };
    };
  },
  /**
   * weight
   */
  number,
];

export type XTokensTransferMultiCurrenciesExtrinsicParams<
  Symbols extends AssetSymbol = AssetSymbol,
> = [
  [
    [
      XTokensTransferExtrinsicParamsAsset<Symbols>,
      /**
       * amount
       */
      bigint,
    ],
    [
      XTokensTransferExtrinsicParamsAsset<Symbols>,
      /**
       * fee
       */
      bigint,
    ],
  ],
  /**
   * fee item
   */
  1,
  {
    V1: {
      parents: 1;
      interior: {
        X2: [
          {
            Parachain: number;
          },
          {
            AccountKey20: {
              network: 'Any';
              /**
               * account
               */
              key: string;
            };
          },
        ];
      };
    };
  },
  /**
   * weight
   */
  number,
];

export type XTokensTransferExtrinsicParamsAsset<
  Symbols extends AssetSymbol = AssetSymbol,
> =
  | {
      [XTokensExtrinsicCurrencyTypes.Token]: Symbols | 'KUSD' | 'MOVR';
    }
  | {
      [XTokensExtrinsicCurrencyTypes.Token2]: AssetId;
    }
  | {
      [XTokensExtrinsicCurrencyTypes.Native]: Symbols;
    }
  | {
      [XTokensExtrinsicCurrencyTypes.ForeignAsset]: AssetId;
    }
  | {
      [XTokensExtrinsicCurrencyTypes.MantaCurrency]: AssetId;
    }
  | {
      [XTokensExtrinsicCurrencyTypes.OtherReserve]: AssetId;
    }
  | Symbols
  | AssetId
  | 'SelfReserve';
