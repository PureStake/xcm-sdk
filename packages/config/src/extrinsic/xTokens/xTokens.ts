import { Asset } from '../../constants';
import { AssetConfig, ChainConfig, MoonChainConfig } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  XTokensExtrinsic,
  XTokensExtrinsicSuccessEvent,
} from './xTokens.constants';
import {
  XTokensTransferExtrinsic,
  XTokensTransferExtrinsicParamsAsset,
  XTokensTransferMultiAssetExtrinsic,
  XTokensTransferMultiCurrenciesExtrinsic,
} from './xTokens.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function xTokens<Assets extends Asset>(chain: MoonChainConfig) {
  return {
    transfer: () => transfer<Assets>(chain),
    transferMultiAsset: () => transferMultiAsset<Assets>(chain),
    transferMultiCurrencies: () => transferMultiCurrencies<Assets>(chain),
  };
}

function transfer<Assets extends Asset>(chain: MoonChainConfig) {
  return {
    successEvent: (event: XTokensExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => ({
        asset: (
          token: XTokensTransferExtrinsicParamsAsset<Assets>,
        ): XTokensTransferExtrinsic<Assets> => ({
          pallet: ExtrinsicPallet.XTokens,
          extrinsic: XTokensExtrinsic.Transfer,
          successEvent: event,
          getParams: (account, amount) => [
            token,
            amount,
            {
              V1: {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: chain.parachainId,
                    },
                    {
                      AccountKey20: {
                        network: 'Any',
                        key: account,
                      },
                    },
                  ],
                },
              },
            },
            origin.weight,
          ],
        }),
      }),
    }),
  };
}

function transferMultiAsset<Assets extends Asset>(chain: MoonChainConfig) {
  return {
    successEvent: (event: XTokensExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => ({
        asset: (
          asset: AssetConfig<Assets>,
        ): XTokensTransferMultiAssetExtrinsic<Assets> => ({
          pallet: ExtrinsicPallet.XTokens,
          extrinsic: XTokensExtrinsic.TransferMultiAsset,
          successEvent: event,
          getParams: (account, amount) => [
            {
              V1: {
                id: {
                  Concrete: {
                    parents: 1,
                    interior: {
                      X2: [
                        {
                          Parachain: origin.parachainId,
                        },
                        {
                          GeneralKey: asset.originSymbol,
                        },
                      ],
                    },
                  },
                },
                fun: {
                  Fungible: amount,
                },
              },
            },
            {
              V1: {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: chain.parachainId,
                    },
                    {
                      AccountKey20: {
                        network: 'Any',
                        key: account,
                      },
                    },
                  ],
                },
              },
            },
            origin.weight,
          ],
        }),
      }),
    }),
  };
}

function transferMultiCurrencies<Assets extends Asset>(chain: MoonChainConfig) {
  return {
    successEvent: (event: XTokensExtrinsicSuccessEvent) => ({
      origin: (origin: ChainConfig) => ({
        assets: (
          asset: AssetConfig<Assets>,
          feeAsset: AssetConfig<Assets>,
        ): XTokensTransferMultiCurrenciesExtrinsic<Assets> => ({
          pallet: ExtrinsicPallet.XTokens,
          extrinsic: XTokensExtrinsic.TransferMultiCurrencies,
          successEvent: event,
          getParams: (account, amount, fee = 0n) => [
            [
              [
                {
                  Token: asset.originSymbol,
                },
                amount,
              ],
              [
                {
                  Token: feeAsset.originSymbol,
                },
                fee,
              ],
            ],
            1,
            {
              V1: {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: chain.parachainId,
                    },
                    {
                      AccountKey20: {
                        network: 'Any',
                        key: account,
                      },
                    },
                  ],
                },
              },
            },
            origin.weight,
          ],
        }),
      }),
    }),
  };
}
