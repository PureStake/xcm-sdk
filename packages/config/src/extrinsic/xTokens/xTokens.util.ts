import { SubmittableExtrinsicFunction } from '@polkadot/api/types';
import { XTokensWeightLimit } from './xTokens.interfaces';

/**
 * Workaround for weight parameter type change in xTokens pallet
 * https://github.com/open-web3-stack/open-runtime-module-library/pull/841
 */
export function getXTokensWeight(
  weight: number,
  extrinsicCall?: SubmittableExtrinsicFunction<'promise'>,
): XTokensWeightLimit {
  const type = extrinsicCall?.meta.args.at(-1)?.type;

  if (!type) {
    return weight;
  }

  if (type.eq('XcmV2WeightLimit')) {
    return {
      Limited: weight,
    };
  }

  if (type.eq('XcmV3WeightLimit')) {
    return {
      Limited: { refTime: weight, proofSize: 0 },
    };
  }

  return weight;
}
