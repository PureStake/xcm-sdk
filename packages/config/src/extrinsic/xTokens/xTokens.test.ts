import { AssetSymbol, ChainKey, MoonChainKey } from '../../constants';
import { Chain, MoonChain } from '../../interfaces';
import { xTokens } from './xTokens';
import {
  XTokensExtrinsicCurrencyTypes,
  XTokensExtrinsicSuccessEvent,
} from './xTokens.constants';

describe('xTokens', () => {
  const amount = 1000n;
  const account = '0xeF46c7649270C912704fB09B75097f6E32208b85';
  const chain: MoonChain = {
    key: MoonChainKey.MoonbaseAlpha,
    name: 'Moonbase Alpha',
    ws: 'wss://wss.api.moonbase.moonbeam.network',
    parachainId: 1000,
    decimals: 10,
    chainId: 1287,
    unitsPerSecond: 100n,
  };
  const origin: Chain = {
    key: ChainKey.AlphanetRelay,
    name: 'Alphanet Relay',
    ws: 'wss://frag-moonbase-relay-rpc-ws.g.moonbase.moonbeam.network',
    weight: 1_000_000_000,
    parachainId: 0,
  };
  const asset = {
    id: '16797826370226091782818345603793389938',
    erc20Id: '0xffffffff0ca324c842330521525e7de111f38972',
    originSymbol: AssetSymbol.ASTR,
  };
  const extrinsic = xTokens(chain);

  describe('transfer', () => {
    const cfg = extrinsic
      .transfer()
      .successEvent(XTokensExtrinsicSuccessEvent.Transferred)
      .origin(origin)
      .asset({ Token: AssetSymbol.MOVR });

    it('should be correct config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should get correct params', () => {
      expect(cfg.getParams(account, amount)).toMatchSnapshot();
    });
  });

  describe('transferMultiAsset', () => {
    const cfg = extrinsic
      .transferMultiAsset()
      .successEvent(XTokensExtrinsicSuccessEvent.Transferred)
      .origin(origin)
      .asset(asset);

    it('should be correct config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should get correct params', () => {
      expect(cfg.getParams(account, amount)).toMatchSnapshot();
    });
  });

  describe('transferMultiCurrencies', () => {
    const cfg = extrinsic
      .transferMultiCurrencies()
      .successEvent(XTokensExtrinsicSuccessEvent.Transferred)
      .origin(origin)
      .assets(
        {
          [XTokensExtrinsicCurrencyTypes.Token]: asset.originSymbol,
        },
        {
          [XTokensExtrinsicCurrencyTypes.Token]: asset.originSymbol,
        },
      );

    it('should be correct config', () => {
      expect(cfg).toMatchSnapshot();
    });

    it('should get correct params', () => {
      expect(cfg.getParams(account, amount, 99n)).toMatchSnapshot();
    });
  });
});
