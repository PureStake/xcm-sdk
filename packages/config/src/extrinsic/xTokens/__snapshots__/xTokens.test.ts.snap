// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xTokens transfer should be correct config 1`] = `
{
  "extrinsic": "transfer",
  "getParams": [Function],
  "pallet": "xTokens",
  "successEvent": "Transferred",
}
`;

exports[`xTokens transfer should get correct params 1`] = `
[
  {
    "Token": "MOVR",
  },
  1000n,
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
              "network": "Any",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  1000000000,
]
`;

exports[`xTokens transferMultiAsset should be correct config 1`] = `
{
  "extrinsic": "transferMultiasset",
  "getParams": [Function],
  "pallet": "xTokens",
  "successEvent": "Transferred",
}
`;

exports[`xTokens transferMultiAsset should get correct params 1`] = `
[
  {
    "V1": {
      "fun": {
        "Fungible": 1000n,
      },
      "id": {
        "Concrete": {
          "interior": {
            "X2": [
              {
                "Parachain": 0,
              },
              {
                "GeneralKey": "ASTR",
              },
            ],
          },
          "parents": 1,
        },
      },
    },
  },
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
              "network": "Any",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  1000000000,
]
`;

exports[`xTokens transferMultiCurrencies should be correct config 1`] = `
{
  "extrinsic": "transferMulticurrencies",
  "getParams": [Function],
  "pallet": "xTokens",
  "successEvent": "Transferred",
}
`;

exports[`xTokens transferMultiCurrencies should get correct params 1`] = `
[
  [
    [
      {
        "Token": "ASTR",
      },
      1000n,
    ],
    [
      {
        "Token": "ASTR",
      },
      99n,
    ],
  ],
  1,
  {
    "V1": {
      "interior": {
        "X2": [
          {
            "Parachain": 1000,
          },
          {
            "AccountKey20": {
              "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
              "network": "Any",
            },
          },
        ],
      },
      "parents": 1,
    },
  },
  1000000000,
]
`;
